File: internal/ipconfig/domain/stat.go
package domain

import "math"

// 对于gateway网关机来说，存在不同时期加入进来的物理机，所以机器的配置是不同的，使用负载来衡量会导致偏差。
// 为更好的应对动态的机器配置变化，我们统计其剩余资源值，来衡量一个机器其是否更适合增加其负载。
// 这里的数值代表的是，此endpoint对应的机器其，自身剩余的资源指标。
type Stat struct {
	ConnectNum   float64 // 业务上，im gateway 总体持有的长连接数量 的剩余值
	MessageBytes float64 // 业务上，im gateway 每秒收发消息的总字节数 的剩余值
}

// 因此假设网络带宽将是系统瓶颈所在，那么哪台机器富余的带宽资源多，哪台机器的负载就是最轻的。
func (s *Stat) CalculateActiveScore() float64 {
	return getGB(s.MessageBytes)
}

func (s *Stat) CalculateStaticScore() float64 {
	return s.ConnectNum
}

func (s *Stat) Avg(num float64) {
	s.ConnectNum /= num
	s.MessageBytes /= num
}

func (s *Stat) Clone() *Stat {
	newStat := &Stat{
		MessageBytes: s.MessageBytes,
		ConnectNum:   s.ConnectNum,
	}
	return newStat
}

func (s *Stat) Add(st *Stat) {
	if st == nil {
		return
	}
	s.ConnectNum += st.ConnectNum
	s.MessageBytes += st.MessageBytes
}

func (s *Stat) Sub(st *Stat) {
	if st == nil {
		return
	}
	s.ConnectNum -= st.ConnectNum
	s.MessageBytes -= st.MessageBytes
}

func getGB(m float64) float64 {
	return roundToTwoDecimals(m / (1 << 30))
}

func roundToTwoDecimals(value float64) float64 {
	return math.Trunc(value*1e2+0.5) * 1e-2
}

File: im-ipconfig/src/domain/stat.rs
#[derive(Default, Clone, Debug)]
pub(crate) struct Stat {
    pub(crate) connect_num: f64,   // 总体持有的长连接数量 的剩余值
    pub(crate) message_bytes: f64, // 每秒收发消息的总字节数 的剩余值
}

impl Stat {
    #[allow(dead_code)]
    pub(crate) fn new(connect_num: f64, message_bytes: f64) -> Self {
        Stat {
            connect_num,
            message_bytes,
        }
    }

    pub(crate) fn active_score(&self) -> f64 {
        convert_to_gb(self.message_bytes)
    }

    pub(crate) fn static_score(&self) -> f64 {
        self.connect_num
    }

    pub(crate) fn avg(&mut self, num: f64) {
        self.connect_num /= num;
        self.message_bytes /= num;
    }

    pub(crate) fn add(&mut self, other: &Stat) {
        self.connect_num += other.connect_num;
        self.message_bytes += other.message_bytes;
    }

    pub(crate) fn sub(&mut self, other: &Stat) {
        self.connect_num -= other.connect_num;
        self.message_bytes -= other.message_bytes;
    }
}

fn convert_to_gb(bytes: f64) -> f64 {
    round_to_two_decimals(bytes / f64::from(1 << 30))
}

fn round_to_two_decimals(value: f64) -> f64 {
    (value * 100.0).round() / 100.0
}

#[cfg(test)]
mod tests {
    use approx::assert_relative_eq;

    use super::*;

    #[test]
    fn test_calculate_active_score() {
        let stat = Stat::new(100.0, 5_000_000_000.0); // 5 GB
        assert_relative_eq!(stat.active_score(), 4.66); // 5 GB in GB, rounded to two decimal places
    }

    #[test]
    fn test_calculate_static_score() {
        let stat = Stat::new(100.0, 5_000_000_000.0);
        assert_relative_eq!(stat.static_score(), 100.0);
    }

    #[test]
    fn test_avg() {
        let mut stat = Stat::new(100.0, 5_000_000_000.0);
        stat.avg(2.0);
        assert_relative_eq!(stat.connect_num, 50.0);
        assert_relative_eq!(stat.message_bytes, 2_500_000_000.0);
    }

    #[test]
    fn test_add() {
        let mut stat1 = Stat::new(100.0, 5_000_000_000.0);
        let stat2 = Stat::new(50.0, 2_500_000_000.0);
        stat1.add(&stat2);
        assert_relative_eq!(stat1.connect_num, 150.0);
        assert_relative_eq!(stat1.message_bytes, 7_500_000_000.0);
    }

    #[test]
    fn test_sub() {
        let mut stat1 = Stat::new(100.0, 5_000_000_000.0);
        let stat2 = Stat::new(50.0, 2_500_000_000.0);
        stat1.sub(&stat2);
        assert_relative_eq!(stat1.connect_num, 50.0);
        assert_relative_eq!(stat1.message_bytes, 2_500_000_000.0);
    }
}
